[Osx] native windowing/input implementation

fuzzard opened the pull request:
## Description
This is the first steps to a "native" windowing and input handling without the requirement for SDL on OSX.

## Motivation and context
SDL1 hasnt been updated for years, and i dont believe its worth the time to migrate to SDL2 for a single platform that has had previous work to remove completely.
The main push for this was to support Apple Silicon builds, and not stuff around with SDL.

The bulk of this work is pulled from the existing osxremovesdl branch, and has been adapted to match all the core changes that have gone in since.

My intention is to try and get this mainlined, in the hopes that we may get some more devs interested in working through the edge cases and quirks if a base "working" implementation is there.

Ive pinged @Memphiz in case he finds any time for any input/review. No dramas if you dont have any time
Theres absolutely edge cases that arent handled (multiple screens untested/unimplemented, Full screen needs some work, etc), so the default build will remain defaulting to SDL. Ive looked to extract the SDL specific code into subfolders so when the time comes and native windowing becomes the preferred, the removal of SDL should be quite easy.

## How has this been tested?
OSX Intel based macbook

## What is the effect on users?
None right now. Default windowing still defaults to SDL (ie all jenkins builds/releases)

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Commit:
[OSX] native input handling implementation

Commit:
[OSX] native windowing implementation

Commit:
[osx] relocate more SDL specific code

garbear commented:
I think it's worth deciding: Do we make SDL removal a feature of Nexus? I say we pull the trigger and never look back, as there's no pressure to rush Nexas atm. In that case, better to merge "good enough" as soon as possible to move the effort toward completion for Nexus.

garbear commented on xbmc/platform/darwin/osx/SDL/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+if(SDL_FOUND)
```
I would remove the `if` statement here. Anything in `xbmc/platform/darwin/osx/SDL` must be ignored if SDL isn't found. That means the `if(SOURCES)` can also be removed, right?

fuzzard replied:
The issue is our cmake treedata cant handle states like that, its just all pulled in and folders parsed for CMakeLists.txt files.

The other option would be to not add xbmc/platform/darwin/osx/SDL to the cmake/treedata/osx/subdirs.txt and do the inclusion behind the same guard in xbmc/platform/darwin/osx/CMakeLists.txt

garbear replied:
OK, I think it's fine here

garbear commented:
Can I ask about joysticks and the Cocoa driver in peripheral.joystick? Does that change for new macs?

fuzzard commented:
> I think it's worth deciding: Do we make SDL removal a feature of Nexus? I say we pull the trigger and never look back, as there's no pressure to rush Nexas atm. In that case, better to merge "good enough" as soon as possible to move the effort toward completion for Nexus.

Short answer is No, i dont believe we would be best to straight drop SDL at this point in time. The native windowing isnt feature compatible with the existing SDL implementation, and i cant guarantee i'll have the time (or desire) to make sure there wouldnt be any regressions.
My hope is that theres enough in here, that users who have interest in all the quirks (Airplay, multiple displays, whatever input testing, multiple gfx cards, etc) may be able to work on small aspects that interest them to provide fixes for stuff i just cant test/replicate/care about.

> Can I ask about joysticks and the Cocoa driver in peripheral.joystick? Does that change for new macs?

In theory we support osx 10.13 as a minimum, so we can probably drop the Cocoa driver and rely on the GCController implementation that was opened up to OSX for v20 a while back. Probably just needs more testing/confirmation. Ive only really tested on ios, with fairly minimal OSX testing to confirm buttons that didnt work in cocoa driver did with GCController

fuzzard commented:
Ok, for any potential reviewers. Ive spent the afternoon tidying up, fixing header inclusions, adding some pragma marking, dot style calls, removing obsolete code, and other minors.

If i change anything else from here, it'll be functional changes, so i'll keep any further changes to fixups/other commits for further review

enen92 commented:
I assume there is no way we can generate a testbuild with jenkins? My mac is getting too old compiling kodi is PITA

kambala-decapitator commented:
> I assume there is no way we can generate a testbuild with jenkins? My mac is getting too old compiling kodi is PITA

@enen92 you can build from any branch on Jenkins, see https://jenkins.kodi.tv/view/Automation/job/BuildMulti-Choice/

fuzzard commented:
Testbuild is available here http://mirrors.kodi.tv/test-builds/osx/x86_64/kodi-20211001-cf27f5a8-PR20200-merge-x86_64.dmg
was built from jenkins job https://jenkins.kodi.tv/job/OSX-64/22577/

Ive added a temp commit that will build native on jenkins. Ill leave that in so team members can trigger a jenkins build if they need to.
I'll continue to leave the "No Jenkins" label on this, as i dont really need/want it clogging up jenkins build queues at this time, so just manually trigger a build if you need it.

enen92 commented:
thanks fuzzard I'll give it a whirl

kambala-decapitator reviewed (changes requested):
I also suggest to run Clang Analyzer on the new files

Commit:
[OSX] Further improvements to windowing/input

relocate GL based code to own folders
fixups for input handling
rework/tidy of XBMCApplication

kambala-decapitator commented on xbmc/windowing/osx/WinSystemOSX.mm:
```diff
@@ -0,0 +1,1370 @@
+
+  // for native fullscreen we always want to set the
+  // same windowed flags
+  __block NSUInteger windowStyleMask;
```
`__block` is only needed when:

- variable is assigned a value in a block
- you want to capture variable by reference in a block (default is to capture by value)

iirc none of these apply to this variable

kambala-decapitator commented on xbmc/platform/darwin/osx/XBMCApplication.mm:
```diff
@@ -134,15 +116,10 @@ @implementation XBMCDelegate
-  }
-  CFRelease(url);
-  CFRelease(url2);
+  NSString* bundlePath = [[NSBundle mainBundle] resourcePath];
+  NSString* mainPath = [bundlePath stringByDeletingLastPathComponent];
+
+  assert(chdir(mainPath.UTF8String) == 0); /* chdir to the binary app's parent */
```
If Kodi is in /Applications/Kodi.app, original code would point to /Applications
```suggestion
  auto parentPath = NSBundle.mainBundle.bundlePath.stringByDeletingLastPathComponent;
  NSAssert([NSFileManager.defaultManager changeCurrentDirectoryPath:parentPath]);
```

I see that this code was copied from the old implementation, but which purpose it serves? Is it still required?

fuzzard replied:
RE the required state, Id honestly prefer to just leave it as is for now. Dont want to go chasing shit when theres already so much other stuff happening in this PR. Ive no idea what its used for, or why its there, So i'll add a todo to check if its still relevant at a later point in time

kambala-decapitator replied:
ok, let's keep it, just apply the suggestion please

kambala-decapitator commented on xbmc/platform/darwin/osx/XBMCApplication.mm:
```diff
@@ -383,14 +358,48 @@ - (void)floatOnTopToggle:(id)sender
+  menuItem = [[NSMenuItem alloc] initWithTitle:@"Float on Top"
+                                        action:@selector(floatOnTopToggle:)
+                                 keyEquivalent:@"t"];
+  menuItem.keyEquivalentModifierMask = NSEventModifierFlagCommand | NSEventModifierFlagControl;
```
existing shortcut is Cmd+T, drop this line

kambala-decapitator commented on xbmc/platform/darwin/osx/XBMCApplication.mm:
```diff
@@ -383,14 +358,48 @@ - (void)floatOnTopToggle:(id)sender
+  menuItem = [[NSMenuItem alloc] initWithTitle:@"Minimize"
+                                        action:@selector(performMiniaturize:)
+                                 keyEquivalent:@"m"];
+  menuItem.keyEquivalentModifierMask = NSEventModifierFlagCommand | NSEventModifierFlagControl;
```
existing shortcut (and macOS default) is Cmd+M, drop this line

kambala-decapitator commented on xbmc/platform/darwin/osx/XBMCApplication.mm:
```diff
@@ -420,25 +429,20 @@ int main(int argc, char* argv[])
-    [NSApplication.sharedApplication setDelegate:xbmc_delegate];
-
+    // Set App Delegate
+    [NSApp setDelegate:[XBMCDelegate new]];
```
delegate will die after this line because the property is `weak`.

```suggestion
    auto appDelegate = [XBMCDelegate new];
    NSApp.delegate = appDelegate;
```

kambala-decapitator commented on xbmc/windowing/osx/WinEventsOSXImpl.mm:
```diff
@@ -90,19 +93,19 @@ - (unichar)OsxKey2XbmcKey:(unichar)character
   {
-    case 0x1c:
-    case 0xf702:
+    case 0x1c: // kVK_ANSI_8
```
import `Events.h` from Carbon framework, the constants are there. https://stackoverflow.com/a/4642095/1971301

kambala-decapitator commented on xbmc/windowing/osx/WinEventsOSXImpl.mm:
```diff
@@ -341,4 +321,28 @@ - (NSEvent*)InputEventHandler:(NSEvent*)nsevent
+  UniCharCount actualStringLength = 0;
+
+  CGKeyCode keycode = (CGKeyCode)CGEventGetIntegerValueField(*event, kCGKeyboardEventKeycode);
+  CGEventKeyboardGetUnicodeString(*event, MAXUNICODESTRLEN,
```
instead of hardcoding max amount it's probably better to make 2 calls to `CGEventKeyboardGetUnicodeString` where first would return the real length. From the header:

> When you call this function with a NULL string or a maximum string length of 0, the function still returns the actual count of Unicode characters in the event.

Although we care only about the first character, then even length 2 should be sufficient.

kambala-decapitator commented on xbmc/windowing/osx/WinEventsOSXImpl.mm:
```diff
@@ -341,4 +321,28 @@ - (NSEvent*)InputEventHandler:(NSEvent*)nsevent
+
+  XBMC_Event newEvent = {};
+
+  if (actualStringLength == 0)
```
it looks to me that you extracted the original code incorrectly. The following parts are missing:

```objc
// 1. when actualStringLength is 0
unicodeString[0] = [self OsxKey2XbmcKey:[nsevent.characters characterAtIndex:0]];

// 2
if (actualStringLength > 1)
  newEvent.key.keysym.unicode |= (unicodeString[1] << 8);
```

fuzzard replied:
Honestly I played with it and couldn't get a string longer than a single character. I've seen other snippets who just stick to a single char, really not sure which way to go

kambala-decapitator replied:
ok, let's stick to a single character then. Maybe add a short comment in case we'll have to adjust it in the future.

kambala-decapitator commented on xbmc/windowing/osx/WinEventsOSXImpl.mm:
```diff
@@ -341,4 +321,28 @@ - (NSEvent*)InputEventHandler:(NSEvent*)nsevent
+  UniChar unicodeString[MAXUNICODESTRLEN] = {0};
+  UniCharCount actualStringLength = 0;
+
+  CGKeyCode keycode = (CGKeyCode)CGEventGetIntegerValueField(*event, kCGKeyboardEventKeycode);
```
```suggestion
  auto keycode = static_cast<CGKeyCode>(CGEventGetIntegerValueField(*event, kCGKeyboardEventKeycode));
```

kambala-decapitator commented on xbmc/windowing/osx/WinSystemOSX.mm:
```diff
@@ -316,13 +319,14 @@ CGDisplayModeRef BestMatchForMode(
     }
   }
 
+  CGDisplayModeRelease(displayMode); // release the copy we got before ...
```
releasing at this point is incorrect because you would return garbage.

this code should be carefully reviewed regarding the reference counts including the calling code.

kambala-decapitator reviewed (changes requested):
we should also consider dropping all the custom code to setup the application object in `main` and use standard approach with Info.plist + `NSApplicationMain(argc, argv)` instead (see template when you create a new macOS app in Xcode).

kambala-decapitator commented on xbmc/windowing/osx/WinEventsOSXImpl.mm:
```diff
@@ -0,0 +1,344 @@
+      NSEventMaskLeftMouseDown | NSEventMaskLeftMouseUp | NSEventMaskRightMouseDown |
+      NSEventMaskRightMouseUp | NSEventMaskLeftMouseDragged | NSEventMaskRightMouseDragged |
+      NSEventMaskOtherMouseDown | NSEventMaskOtherMouseUp | NSEventMaskOtherMouseDragged |
+      NSEventMaskMouseMoved | NSEventMaskScrollWheel | NSEventMaskKeyDown | NSEventMaskKeyUp;
```
thanks, looks much better!

maybe drags can be grouped on a single line, but that's up to you.

fuzzard commented on xbmc/input/XBMC_keysym.h:
```diff
@@ -27,6 +27,7 @@ typedef enum
   // The keyboard syms have been cleverly chosen to map to ASCII
   XBMCK_UNKNOWN = 0x00,
   XBMCK_FIRST = 0x00,
+  XBMCK_CTRLF = 0x06,
```
https://en.m.wikipedia.org/wiki/List_of_Unicode_characters
But more importantly, it's what's actually returned when skipping through the nsevent of key combo press

fuzzard commented on xbmc/platform/darwin/osx/XBMCApplication.mm:
```diff
@@ -0,0 +1,445 @@
+ *
+ * This message is ignored once the app's mainline has been called.
+ */
+- (BOOL)application:(NSApplication*)theApplication openFile:(NSString*)filename
```
Have added a todo. Will leave this for another PR

fuzzard commented on xbmc/windowing/osx/WinSystemOSX.mm:
```diff
@@ -0,0 +1,1370 @@
+static NSWindow* curtainWindow;
+void fadeInDisplay(NSScreen* theScreen, double fadeTime)
+{
+  int fadeSteps = 100;
+  double fadeInterval = (fadeTime / (double)fadeSteps);
+
+  if (curtainWindow != nil)
+  {
+    for (int step = 0; step < fadeSteps; step++)
+    {
+      double fade = 1.0 - (step * fadeInterval);
+      [curtainWindow setAlphaValue:fade];
+
+      NSDate* nextDate = [NSDate dateWithTimeIntervalSinceNow:fadeInterval];
+      [NSRunLoop.currentRunLoop runUntilDate:nextDate];
+    }
+  }
```
Will do this as a todo as well.

kambala-decapitator commented on xbmc/windowing/osx/WinSystemOSX.mm:
```diff
@@ -1293,8 +1293,7 @@ static void DisplayReconfigured(CGDirectDisplayID display,
 {
-  std::unique_ptr<CVideoSync> pVSync(new CVideoSyncOsx(clock));
-  return pVSync;
+  return std::make_unique<CVideoSync>(clock);
```
```suggestion
  return std::make_unique<CVideoSyncOsx>(clock);
```

kambala-decapitator commented on xbmc/windowing/osx/WinSystemOSX.mm:
```diff
@@ -1317,7 +1316,7 @@ static void DisplayReconfigured(CGDirectDisplayID display,
 std::unique_ptr<IOSScreenSaver> CWinSystemOSX::GetOSScreenSaverImpl()
 {
-  return std::unique_ptr<IOSScreenSaver>(new COSScreenSaverOSX);
+  return std::make_unique<IOSScreenSaver>();
```
```suggestion
  return std::make_unique<COSScreenSaverOSX>();
```

kambala-decapitator commented on xbmc/windowing/osx/WinEventsOSXImpl.mm:
```diff
@@ -324,19 +322,33 @@ - (NSEvent*)InputEventHandler:(NSEvent*)nsevent
-  auto keycode = static_cast<CGKeyCode>(CGEventGetIntegerValueField(*event, kCGKeyboardEventKeycode));
-  CGEventKeyboardGetUnicodeString(*event, MAXUNICODESTRLEN,
+  // Create array with size of event string
+  UniChar unicodeString[actualStringLength];
+  memset(unicodeString, 0, sizeof(unicodeString));
```
```suggestion
  UniChar unicodeString[actualStringLength] = {0};
```

fuzzard replied:
Can't default init variable length array. Already tried that, haha. Have to memset

kambala-decapitator commented on xbmc/windowing/osx/WinEventsOSXImpl.mm:
```diff
@@ -324,19 +322,33 @@ - (NSEvent*)InputEventHandler:(NSEvent*)nsevent
+  auto keycode =
+      static_cast<CGKeyCode>(CGEventGetIntegerValueField(*event, kCGKeyboardEventKeycode));
+  CGEventKeyboardGetUnicodeString(*event, sizeof(unicodeString) / sizeof(*unicodeString),
                                   &actualStringLength, unicodeString);
```
```suggestion
  CGEventKeyboardGetUnicodeString(*event, actualStringLength,
                                  nullptr, unicodeString);
```

fuzzard replied:
The nullptr doesnt actually work. events fail to get processed unless the UniCharCount is provided.

kambala-decapitator commented on xbmc/windowing/osx/WinEventsOSXImpl.mm:
```diff
@@ -324,19 +322,33 @@ - (NSEvent*)InputEventHandler:(NSEvent*)nsevent
-  UniChar unicodeString[MAXUNICODESTRLEN] = {0};
   UniCharCount actualStringLength = 0;
+  // Get stringlength of event
+  CGEventKeyboardGetUnicodeString(*event, 0, &actualStringLength, NULL);
```
```suggestion
  CGEventKeyboardGetUnicodeString(*event, 0, &actualStringLength, nullptr);
```

kambala-decapitator commented on xbmc/windowing/osx/WinSystemOSX.mm:
```diff
@@ -0,0 +1,1376 @@
+      double tmpRefresh;
+
+      GetScreenResolution(&tmpWidth, &tmpHeight, &tmpRefresh, screenIdx);
+      dispMode = GetMode(tmpWidth, tmpHeight, tmpRefresh, screenIdx);
```
`CGDisplayModeRelease(dispMode);` must be done before assignment

kambala-decapitator commented on xbmc/windowing/osx/WinSystemOSX.mm:
```diff
@@ -0,0 +1,1376 @@
+
+      // no way to get a resolution set
+      if (!dispMode)
+        return false;
```
you forgot to tell that we have a match now:

```suggestion
        return false;
      match = true;
```

kambala-decapitator commented on xbmc/windowing/osx/WinSystemOSX.mm:
```diff
@@ -0,0 +1,1376 @@
+        (rate == refreshrate || rate == 0))
+    {
+      CLog::Log(LOGDEBUG, "GetMode found a match!");
+      return displayMode;
```
you forgot to release the array:

```suggestion
      CFRelease(displayModes);
      return displayMode;
```

fuzzard replied:
Have gone through and checked all the cf*ref usage. I believe everything is released/retained where appropriate now

lrusak commented on xbmc/windowing/osx/WinEventsOSXImpl.mm:
```diff
@@ -0,0 +1,365 @@
+  CGEventKeyboardGetUnicodeString(*event, 0, &actualStringLength, NULL);
+
+  // Create array with size of event string
+  UniChar unicodeString[actualStringLength];
```
What about using std::vector instead?

fuzzard replied:
Yeah, its just not worth it. UniChar type is expected for when its used. Casting in and out just isnt worth it imo

kambala-decapitator replied:
you don't need to cast vector, simply pass `vec.data()`

lrusak commented on xbmc/windowing/osx/WinEventsOSXImpl.mm:
```diff
@@ -0,0 +1,365 @@
+  if (appleModifier & kCGEventFlagMaskCommand)
+    xbmcModifier |= XBMCKMOD_LMETA;
+
+  return (XBMCMod)xbmcModifier;
```
c++ style cast

lrusak commented on xbmc/windowing/osx/WinEventsOSXImpl.mm:
```diff
@@ -0,0 +1,365 @@
+  unicodeString[0] = [self OsxKey2XbmcKey:unicodeString[0]];
+
+  newEvent.key.keysym.scancode = keycode;
+  newEvent.key.keysym.sym = (XBMCKey)unicodeString[0];
```
c++ style cast

lrusak commented on xbmc/platform/darwin/osx/XBMCApplication.mm:
```diff
@@ -0,0 +1,390 @@
+  CServiceBroker::GetLogging().SetLogLevel(LOG_LEVEL_DEBUG);
+#else
+  //  CServiceBroker::GetLogging().SetLogLevel(LOG_LEVEL_NORMAL);
+  CServiceBroker::GetLogging().SetLogLevel(LOG_LEVEL_DEBUG);
```
Is this intentional?

fuzzard replied:
Honestly, no idea. Carried over from existing master. It'll be another one of those, dont want to touch in this pr things. Can look at in a followup

lrusak commented on xbmc/platform/darwin/osx/XBMCApplication.mm:
```diff
@@ -0,0 +1,390 @@
+  if (arg == NULL)
+    return FALSE;
+
+  newargv = (char**)realloc(gArgv, sizeof(char*) * (gArgc + 2));
```
This whole method looks like it could use some love but this looks like a great use for std::vector instead of malloc/realloc

fuzzard replied:
Its already flagged for a todo to replace with a different method. Will leave it as is for now, as it'll be removed completely and replaced with an alternative function (at some point).

fuzzard commented:
Let me know if you spot anymore ref releases ive missed. think its all covered now. If this looks good, i'll rebase all the fixups into a single fixup for any future reviews after this point

lrusak commented on xbmc/Application.cpp:
```diff
@@ -4114,13 +4114,15 @@ void CApplication::ProcessSlow()
   CServiceBroker::GetPowerManager().ProcessEvents();
 
 #if defined(TARGET_DARWIN_OSX)
+#if defined(SDL_FOUND)
```
#if defined(A) && defined(B)

lrusak reviewed (commented):
Just a couple things I noticed 😃

fuzzard commented:
```
we should also consider dropping all the custom code to setup the application object in main and use standard approach with Info.plist + NSApplicationMain(argc, argv) instead (see template when you create a new macOS app in Xcode).
```

RE the above kambala, my understanding is NSApplicationMain only really does loading from a nib/storyboard. At this point, as we dont use a nib/storyboard at all, i dont think it provides any benefit to switch over to it. If we introduce a nib/storyboard and dont manually handle all the menus stuff and whatnot, then we can go that route.
As it stands, another one of the long todo items for another time.

kambala-decapitator commented:
> Let me know if you spot anymore ref releases ive missed. think its all covered now. If this looks good, i'll rebase all the fixups into a single fixup for any future reviews after this point

Seems good so far, but I suggest to run Clang Analyzer.

> > we should also consider dropping all the custom code to setup the application object in main and use standard approach with Info.plist + NSApplicationMain(argc, argv) instead (see template when you create a new macOS app in Xcode).
> 
> RE the above kambala, my understanding is NSApplicationMain only really does loading from a nib/storyboard. At this point, as we dont use a nib/storyboard at all, i dont think it provides any benefit to switch over to it. If we introduce a nib/storyboard and dont manually handle all the menus stuff and whatnot, then we can go that route. As it stands, another one of the long todo items for another time.

nib/storyboard would be good IMO - less code for static stuff that doesn't change a lot over years.

fuzzard commented:
added a new commit that seems to fix the issues with transitioning from window -> full -> window.
Simplified SetFullScreen.

Todos: (a follow up PR one day)

blanking
fade
retain window positioning (not sure if its needed)

fuzzard commented:
I fired off a testbuild of the current state

http://mirrors.kodi.tv/test-builds/osx/x86_64/kodi-20211102-ee706edb-PR20200-merge-x86_64.dmg

If anyone wants to give any feedback

kambala-decapitator reviewed (approved):
code seems fine in general, will try to use the test build over the next days

kambala-decapitator commented:
launched for the first time, this doesn't look right:

![Screenshot 2021-11-10 at 23 56 18](https://user-images.githubusercontent.com/1557784/141192443-76bdeb01-6150-4440-ac69-fc557db24108.png)

- dock and menubar don't disappear, menubar cuts the top portion of the app (the library refresh message is cut).
- when app is launched, it first jumps to another space as if using native fullscreen mode, but then moves back to the main one
- my current display setting says to launch the app on another monitor, but it is off now
- in the log I see that correct display mode is chosen: `GUI format 2560x1440, Display BenQ GW2765 (1952342468): 2560x1440 @ 59.88Hz`

fuzzard commented:
I'll be honest, haven't had the ability to test anything outside of a single screen MacBook config. Anything dual screen is anyone's guess of what will happen, haha.

kambala-decapitator commented:
played a bit with different video modes.

> dock and menubar don't disappear, menubar cuts the top portion of the app (the library refresh message is cut).

this happens on app launch regardless of the selected monitor in settings.

If you toggle window mode after launch, you end up with this (still in a native fullscreen):

![Screenshot 2021-11-11 at 10 32 58](https://user-images.githubusercontent.com/1557784/141257350-aec491cd-9ed0-4ae7-9e50-db2735f7f728.png)

After switching monitor to default or current one in display settings, dock and menubar disappear and behave correctly. However, system mouse cursor doesn't appear when you hover over menubar. Windowed mode also starts working correctly, but window has no border/title bar which makes it impossible to drag around.

I can record a video in case my description is unclear.

kambala-decapitator commented:
made one more test: open app, switch to windowed, exit. Now on launch app window has title bar and I can move it around.

Then I switched back to fullscreen and exited. On next launch fullscreen looks like this and I can no longer fix it with any manipulations:

![Screenshot 2021-11-11 at 21 34 57](https://user-images.githubusercontent.com/1557784/141351525-aa7c5a11-5779-4af7-bd2d-35f25bbe6179.png)

I think some incorrect values were written to `~/Library/Preferences/org.xbmc.kodi.plist`:

    "NSWindow Frame OSXGLWindowPositionHeightWidth" = "671 320 1176 779 0 0 2560 1415 ";

jokala commented:
> played a bit with different video modes.
> 
> > dock and menubar don't disappear, menubar cuts the top portion of the app (the library refresh message is cut).
> 
> this happens on app launch regardless of the selected monitor in settings.
> 
> If you toggle window mode after launch, you end up with this (still in a native fullscreen):
> 
> ![Screenshot 2021-11-11 at 10 32 58](https://user-images.githubusercontent.com/1557784/141257350-aec491cd-9ed0-4ae7-9e50-db2735f7f728.png)
> 
> After switching monitor to default or current one in display settings, dock and menubar disappear and behave correctly. However, system mouse cursor doesn't appear when you hover over menubar. Windowed mode also starts working correctly, but window has no border/title bar which makes it impossible to drag around.
> 
> I can record a video in case my description is unclear.

Setting - System - Display mode - Display 1920x1080p - scroll down, I saw two 1920x1080p, I choose the other 1920x1080p, finally full screen,it works for me.

kambala-decapitator commented:
doesn't switch to the second display, resolution changes on the active one

kambala-decapitator commented:
@fuzzard could you add `NSHighResolutionCapable=true` to `Info.plist` and set GL view's `wantsBestResolutionOpenGLSurface` to `YES` and see if "it just works"?

fuzzard commented:
squashed all the fixups into single review commit

markst commented:
@fuzzard thanks for the work on this! is there any update?

fuzzard commented:
Win32 test failure completely unrelated.

gmonteir commented:
Not sure if I should make a separate issue for this, but M1 builds currently just shows a black screen as of the latest commit. Is there a quick fix for this?

fuzzard commented:
No, honestly you'll need to dig into it yourself at this point in time.

Would be better to open a forum thread, and try and draw help from the community there.
